### PR TITLE
Update note picker to always pick the most notes

### LINKED ISSUE #55: Change note picking to pick the maximum number of notes
**Describe what you want implemented**
Currently the notes are picked using a "minimum number of notes" strategy, by sorting the notes by value and picking the largest valued note. If the transaction value is not reached, we pick the second largest note as well. We then repeat this process up to a maximum of four notes.

This process should be changed to a "maximum number of notes" one. The largest possible grouping - up to a maximum of four -  of notes with the smallest value should be picked - if there is a tie between groups, one should choose the one with more notes.

**Describe "Why" this is needed**
The current algorithm minimizes the number of notes used. While this effectively minimizes the gas cost of a transaction it also maximizes the number of notes produced in a transaction. This exacerbates the so called note fragmentation problem.

The proposed implementation would mitigate the problem by maximizing the number of notes used, and as such minimizing the number of notes produced.

**Describe alternatives you've considered**
N/A

**Additional context**
In testnet users encountered this problem far sooner than expected by continuously transferring to themselves, effectively increasing the fragmentation of their notes with each transaction.

### PR DESCRIPTION
**Pull request by ureeves**
Performing a transaction usually generates two output notes - one note
intended for the target of the transaction, and a change note for the
transactor. When numerous transactions are performed, the previous
algorithm leads to a proliferation of small change notes in a user's
wallet exacerbating the fragmentation issue, eventually preventing a
user from transacting larger amounts.

This algorithm is an effort to mitigate the fragmentation issue, by
always picking the maximum number of notes available and allowed, and
minimizing usage of large notes. This leads to to the desired asymptotic behavior
of minimizing the number of notes in a user's wallet.

Resolves #55

### CODE REVIEW

**Comment by ureeves** [comment]
One, not-insignificant, drawback of doing this is that this note-picking strategy will maximize the amount of gas used per transaction. Maybe this could be critically analyzed in the future to come up with a kind of "middle-of-the-road" approach.

**Comment by autholykos** [comment]
Personally I think that the upcoming plans to migrate to the VM2 mitigate any real problem with overall gas expenditures, although in this case I don't expect the ZK primitives to really be ameliorated with the use of RKYV

File: src/imp.rs [hunk 1]
```diff
@@ -222,27 +222,10 @@ where
...
-            accumulated_value += val;
-            inputs.push((note, val, blinder));
-        }
-
-        if inputs.len() > MAX_INPUT_NOTES {
-            return Err(Error::NoteCombinationProblem);
-        }
-
-        let change = accumulated_value - value;
+        let change = inputs.iter().map(|(_, val, _)| val).sum::<u64>() - value;
```
Comment (ZER0) [inline src/imp.rs:228]: I think you can just `map` as: `map(|v| v.1)` it's shorter and still clear what it does.

File: src/imp.rs [hunk 2]
```diff
@@ -526,6 +509,45 @@ where
...
+/// ascending value order, and go through each contiguous slice of the largest
+/// possible size until we find the first slice whose sum is larger or equal to
+/// the given value. If such a slice is not found, None is returned.
+///
+/// Note: it is presupposed that the input notes contain enough balance to cover
+/// the given `value`.
+fn pick_notes(
+    value: u64,
+    mut notes_and_values: Vec<(Note, u64, JubJubScalar)>,
+) -> Option<Vec<(Note, u64, JubJubScalar)>> {
```
Comment (ZER0) [inline src/imp.rs:525]: I see a couple of API issue here. The most important one is, we're taking a mutable ownership of the `notes_and_values` and we returns the same type; and we have an `Option<Vec<>>`.

And about the mutable ownership, we might want to have a mutable reference and then "reduce" the inputs to the outputs we want – and we remove the return value – or we could have a reference to the input (so, no mutation) and we could return a new value.
Having this signature feels a bit akward IMVHO.

I believe that in this case, an empty vector would be equivalent to `None`, so we might remove the `Option` here and we can just check `is_empty()`.

File: src/imp.rs [hunk 3]
```diff
@@ -526,6 +509,45 @@ where
...
+    let mut picked = Vec::with_capacity(MAX_INPUT_NOTES);
+
+    for i in 0..len - MAX_INPUT_NOTES {
+        if notes_and_values[i..i + MAX_INPUT_NOTES]
+            .iter()
+            .map(|(_, val, _)| val)
+            .sum::<u64>()
+            >= value
+        {
+            picked.extend(notes_and_values.drain(i..i + MAX_INPUT_NOTES));
```
Comment (ZER0) [inline src/imp.rs:543]: Even with this approach I don't think `picked` is actually ever needed. You could just have:
```rust
return Some(
    notes_and_values.drain(i..i + MAX_INPUT_NOTES).collect(),
);
```
(or you could have a local `picked` variable, but not need to preallocate the capacity at this point).

**Review by ZER0 — APPROVED** [review]
LGTM, just a couple of comments maybe to address.

File: src/imp.rs [hunk 4]
```diff
@@ -632,3 +696,81 @@ fn generate_obfuscated_note<Rng: RngCore + CryptoRng>(
...
+
+        let notes_and_values = gen_notes(&values);
+
+        let picked = pick_notes(10, notes_and_values);
+        assert_eq!(picked.len(), 4);
+    }
+
+    #[test]
+    fn note_picking_4_plus() {
+        let values = [10, 20, 30, 40, 50, 60, 70];
```
Comment (t00ts) [inline src/imp.rs:767]: Make it harder for the `pick_notes` to work out the right notes by passing these non sorted

**Review by t00ts — APPROVED** [review]
🚀

**Pull request merged**